watertap/property_models/SOP_prop_pack.py

```diff
@@ -0,0 +1,551 @@
+        )
+
+        def rule_flow_vol_phase_comp(b, p, j):
+            return (
+                b.flow_vol_phase_comp[p, j] * b.dens_mass_phase_comp[p, j]
+                == b.flow_mass_phase_comp[p, j]
+            )
+
+        self.eq_flow_vol_phase_comp = Constraint(
+            self.params.phase_list,
+            self.params.component_list,
+            rule=rule_flow_vol_phase_comp,
+        )
+
+    def _flow_vol_phase(self):
+        self.flow_vol_phase = Var(
+            self.params.phase_list,
+            initialize=1e-3,
+            bounds=(None, None),
+            units=units.m**3 / units.s,
+            doc="Volumetric flow rate",
+        )
+
+        def rule_flow_vol_phase(b, p):
+            return b.flow_vol_phase[p] == sum(
+                b.flow_vol_phase_comp[p, j] for j in b.params.component_list
+            )
+
+        self.eq_flow_vol_phase = Constraint(
+            self.params.phase_list, rule=rule_flow_vol_phase
+        )
+
+    def _flow_mass_phase(self):
+        self.flow_mass_phase = Var(
+            self.params.phase_list,
+            initialize=1e-3,
+            bounds=(None, None),
+            units=units.kg / units.s,
+            doc="Mass flow rate",
+        )
+
+        def rule_flow_mass_phase(b, p):
+            return b.flow_mass_phase[p] == sum(
+                b.flow_mass_phase_comp[p, j] for j in b.params.component_list
+            )
+
+        self.eq_flow_mass_phase = Constraint(
+            self.params.phase_list, rule=rule_flow_mass_phase
+        )
+
+    def _dens_mass_phase(self):
+        self.dens_mass_phase = Var(
+            self.params.phase_list,
+            initialize=1e3,
+            bounds=(1e2, 2e3),
+            units=units.kg * units.m**-3,
+            doc="Mass density",
+        )
+
+        def rule_dens_mass_phase(b, p):
+            return b.dens_mass_phase[p] * b.flow_vol_phase[p] == b.flow_mass_phase[p]
+
+        self.eq_dens_mass_phase = Constraint(
+            self.params.phase_list, rule=rule_dens_mass_phase
+        )
+
+    def _vol_frac_phase_comp(self):
+        self.vol_frac_phase_comp = Var(
+            self.params.phase_list,
+            self.params.component_list,
+            initialize=0.1,
+            bounds=(None, None),
+            units=units.dimensionless,
+            doc="Volumetric fraction",
+        )
+
+        def rule_vol_frac_phase_comp(b, p, j):
+            return (
+                b.vol_frac_phase_comp[p, j] * b.flow_vol_phase[p]
+                == b.flow_vol_phase_comp[p, j]
+            )
+
+        self.eq_vol_frac_phase_comp = Constraint(
+            self.params.phase_list,
+            self.params.component_list,
+            rule=rule_vol_frac_phase_comp,
+        )
+
+    def _conc_mass_phase_comp(self):
+        self.conc_mass_phase_comp = Var(
+            self.params.phase_list,
+            self.params.component_list,
+            initialize=1,
+            bounds=(None, None),
```

Comment:
```suggestion
            domain=NonNegativeReals,
```
another way to write this

Reply:
actually I'm forgetting what the lower bound is for NonNegativeReals (i.e., 0 or None)

Reply:
NonNegativeReals has an effective lower bound of 0 (although it is set separately from bounds in Pyomo). PsotiveReals has effective lower bound of epsilon.

Reply:
I changed the bounds to be `(0, 1)` and the domain to be `NonNegativeReals`. Perhaps that's redundant, but it shouldn't hurt anything.